pkg/controller/clusterpool/clusterpool_controller.go

```diff
@@ -415,6 +423,106 @@ func (r *ReconcileClusterPool) Reconcile(ctx context.Context, request reconcile.
 	return reconcile.Result{}, nil
 }
 
+func (r *ReconcileClusterPool) updateInventory(clp *hivev1.ClusterPool, cds []*hivev1.ClusterDeployment, valid bool, state string, logger log.FieldLogger) {
+	if clp.Spec.Inventory != nil {
+		var active_cdc []string
+		for _, cd := range cds {
+			cdcRef := cd.Spec.ClusterPoolRef.ClusterDeploymentCustomizationRef
+			active_cdc = append(active_cdc, cdcRef.Name)
+			cdc := &hivev1.ClusterDeploymentCustomization{}
+			if err := r.Client.Get(context.Background(), client.ObjectKey{Namespace: clp.Namespace, Name: cdcRef.Name}, cdc); err != nil {
+				if apierrors.IsNotFound(err) {
+					r.setInventoryValidCondition(clp, true, cdc.Name, "missing", logger)
+				}
+				log.WithError(err).Warn("error reading customization")
+				continue
+			}
+			if cdc.Status.ClusterDeploymentRef == nil {
+				cdc.Status.ClusterDeploymentRef = &corev1.ObjectReference{Name: cd.Name, Namespace: cd.Namespace}
+			}
+			cdc.Status.LastApplyTime = metav1.Now()
+			if valid {
+				cdc.Status.LastApplyStatus = "success"
+			} else {
+				cdc.Status.LastApplyStatus = "failed"
+			}
+			conds, changed := controllerutils.SetClusterDeploymentCustomizationCondition(
+				cdc.Status.Conditions,
+				hivev1.ClusterDeploymentCustomizationAvailableCondition,
+				corev1.ConditionFalse,
+				"Reservation",
+				"Reserving cluster deployment customization",
+				controllerutils.UpdateConditionIfReasonOrMessageChange,
+			)
+			if changed {
+				cdc.Status.Conditions = conds
+			}
+
+			if err := r.Status().Update(context.Background(), cdc); err != nil {
+				if apierrors.IsNotFound(err) {
+					r.setInventoryValidCondition(clp, true, cdc.Name, "missing", logger)
+				}
+				log.WithError(err).Warn("failed to update customization status")
+			}
+			r.setInventoryValidCondition(clp, !valid, cdc.Name, state, logger)
+		}
+		sort.Strings(active_cdc)
+		for _, item := range clp.Spec.Inventory {
+			if sort.SearchStrings(active_cdc, item.Name) == len(active_cdc) {
+				cdc := &hivev1.ClusterDeploymentCustomization{}
+				if err := r.Client.Get(context.Background(), client.ObjectKey{Namespace: clp.Namespace, Name: item.Name}, cdc); err != nil {
+					if apierrors.IsNotFound(err) {
+						r.setInventoryValidCondition(clp, true, cdc.Name, "missing", logger)
+					}
+					continue
+				}
+				r.setInventoryValidCondition(clp, false, cdc.Name, "missing", logger)
+				currentAvailability := controllerutils.FindClusterDeploymentCustomizationCondition(
+					cdc.Status.Conditions,
+					hivev1.ClusterDeploymentCustomizationAvailableCondition,
+				)
+				if cdc.Status.ClusterDeploymentRef != nil {
+					cd := &hivev1.ClusterDeployment{}
+					ref := client.ObjectKey{Namespace: cdc.Status.ClusterDeploymentRef.Namespace, Name: cdc.Status.ClusterDeploymentRef.Name}
+					if err := r.Client.Get(context.Background(), ref, cd); err != nil {
```

Comment:
Everywhere (I _think_ without exception) you've added code to `Get()` a ClusterDeployment, you should just be able to use the `cdCollection` cache from [here](https://github.com/openshift/hive/blob/ea63cac7a85032083993849513e53ffe65b3fe83/pkg/controller/clusterpool/clusterpool_controller.go#L293) instead.